src/tree.rs

```diff
@@ -2009,3 +2057,34 @@ impl Debug for Tree {
         Ok(())
     }
 }
+
+/// Compare and swap result.
+///
+/// It returns CAS_OK if operation finishes successfully and otherwise:
+///     - Ok(Err(CompareAndSwapError(cur))) if operation failed to setup a new
+///       value. CompareAndSwapError contains current value.
+///     - Err(Error::Unsupported) if the database is opened in read-only mode.
+pub type CompareAndSwapResult =
+    Result<std::result::Result<(), CompareAndSwapError>>;
+
+/// Successful result of compare and swap operation.
+pub const CAS_OK: CompareAndSwapResult = Ok(Ok(()));
+
+impl From<Error> for CompareAndSwapResult {
+    fn from(error: Error) -> Self {
+        Err(error)
+    }
+}
+
+/// Compare and swap error.
+#[derive(Debug, Clone, PartialEq)]
+pub struct CompareAndSwapError {
+    /// Current value.
+    pub cur: Option<IVec>,
```

Comment:
`cur` -> `current` to be more explicit

I'll add another field that returns the proposed new value that failed to be applied, but don't worry about this for this PR